File: c_lang/15686_ChickenDelivery.c
/***********************************************************/
// [15686] 치킨 배달 (Gold 5)
// 런타임 에러(OutOfBounds)
// 문제에서 집은 최대 100개 치킨집은 최대 13개라 해서
// house, store 배열을 100, 13 + 5 로 줬는데
// 왜 컨테이너 또는 배열에서 할당된 경계를 넘어가는 접근이 발생하지??
// 여유 버퍼 100으로 바꾸면 통과하기는 함...
/***********************************************************/

#include <stdio.h>
#include <stdlib.h>

int N, M;

struct _st {
    int r, c;
};

struct _st house[100 + 100];
struct _st store[13 + 100];
int cnth = 0, cnts = 0;

int dist[13 + 100][100 + 100];
int opened[13 + 100];
int min_dist = 1000000;


void input(void) {
    int tmp;
    scanf("%d %d", &N, &M);
    
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            scanf("%d", &tmp);
            if (tmp == 1) {
                house[cnth].r = i;
                house[cnth].c = j;
                cnth++;
            }
            else if (tmp == 2) {
                store[cnts].r = i;
                store[cnts].c = j;
                cnts++;
            }
        }
    }
}


void dist_table(void) {
    for (int i = 0; i < cnth; i++) {
        for (int j = 0; j < cnts; j++) {
            dist[i][j] = abs(house[i].r - store[j].r) + abs(house[i].c - store[j].c);
        }
    }
}


void combination(int n, int r) {
    int min, sum = 0, tmp;

    if (r == 0) {
        // 선택된 치킨집으로 각 집과의 치킨 거리 게산
        for (int h = 0; h < cnth; h++) {
            min = 1000;
            for (int m = 0; m < M; m++) {
                tmp = dist[h][opened[m]];
                if (tmp < min) min = tmp;
            }
            sum += min;
        }
        
        // 도시의 치킨 거리 최솟값 갱신
        if (sum < min_dist) min_dist = sum;
    }
    else if (n < r) {
        return;
    }
    else {
        opened[r - 1] = n - 1;
        combination(n - 1, r - 1);
        combination(n - 1, r);
    }
}


int main(void) {
    //freopen("in.txt", "r", stdin);
    
    input();
    dist_table();
    combination(cnts, M);
    printf("%d\n", min_dist);

    return 0;
}

File: python/15686_ChickenDelivery.py
# 15686. 치킨 배달 (Gold 5)
'''
* C언어 풀이와의 차이점 
  각각의 집과 치킨집 사이의 거리를 계산한 dist 룩업 테이블을 1차원 배열로 구성

* 시간 단축 연습
1. min() 함수보다 if문으로 직접 비교하는 게 빠르다. 508ms -> 344ms

2. 비교 연산은 정수보다 문자가 느리다. 344ms -> 372ms

3. 전부 입력받고 list comprehension으로 조건에 맞게 필터링 344ms
   C언어처럼 2중 for문으로 입력 받자마자 조건 확인 후 필터링  348ms
'''

from itertools import combinations

N, M = map(int, input().split())

map = [ list(map(int, input().split())) for _ in range(N) ]
map = [ input().split() for _ in range(N) ]

house = [(i, j) for i in range(N) for j in range(N) if map[i][j] == 1]
store = [(i, j) for i in range(N) for j in range(N) if map[i][j] == 2]
cnth, cnts = len(house), len(store)


# C언어처럼 2중 for문으로 입력받자마자 house, store를 구해보자
# house, store = [], []
# cnth, cnts = 0, 0
# for i in range(N):
#     input_ = list(map(int, input().split()))
#     for j, v in enumerate(input_):
#         if v == 1:
#             house.append((i, j))
#             cnth += 1
#         if v == 2:
#             store.append((i, j))
#             cnts += 1

dist = [ abs(h[0]-s[0]) + abs(h[1]-s[1]) for h in house for s in store ]


ans = 1000000
for opened in combinations(range(cnts), M):
    
    min_sum = 0
    for i in range(cnth):
        min_each = 1000
        
        for s in opened:
            tmp = dist[i * cnts + s]
            if tmp < min_each:
                min_each = tmp
            # min_each = min(min_each, dist[i * cnts + s])
        min_sum += min_each
    
    if (min_sum < ans):
        ans = min_sum
    # ans = min(ans, min_sum)
    
print(ans)
